cdc/sorter/leveldb/table_sorter.go

```diff
@@ -0,0 +1,514 @@
+func (ls *Sorter) asyncWrite(
+	ctx context.Context, events []*model.PolymorphicEvent, deleteKeys []message.Key,
+	needIter bool, maxResolvedTs uint64,
+) (chan message.LimitedIterator, error) {
+	// Write and sort events.
+	tk := message.Task{
+		UID:     ls.uid,
+		TableID: ls.tableID,
+		Events:  make(map[message.Key][]byte),
+		IterCh:  make(chan message.LimitedIterator, 1),
+	}
+	for i := range events {
+		event := events[i]
+		if event.RawKV.OpType == model.OpTypeResolved {
+			continue
+		}
+
+		key := encoding.EncodeKey(ls.uid, ls.tableID, event)
+		value := []byte{}
+		var err error
+		value, err = ls.serde.Marshal(event, value)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		tk.Events[message.Key(key)] = value
+	}
+
+	// Delete keys of outputted resolved events.
+	for i := range deleteKeys {
+		tk.Events[deleteKeys[i]] = []byte{}
+	}
+
+	tk.NeedIter = needIter
+	tk.Irange = &lutil.Range{
+		Start: encoding.EncodeTsKey(ls.uid, ls.tableID, 0),
+		Limit: encoding.EncodeTsKey(ls.uid, ls.tableID, maxResolvedTs+1),
+	}
+
+	// Send write task to leveldb.
+	err := ls.router.SendB(ctx, ls.actorID, actormsg.SorterMessage(tk))
+	return tk.IterCh, errors.Trace(err)
+}
+
+// output nonblocking outputs an event. Caller should retry when it returns false.
+func (ls *Sorter) output(event *model.PolymorphicEvent) bool {
+	if ls.lastEvent == nil {
+		ls.lastEvent = event
+	}
+	if ls.lastEvent.CRTs > event.CRTs {
+		log.Panic("regression",
+			zap.Any("lastEntry", ls.lastEvent), zap.Any("event", event),
+			zap.Uint64("regionID", event.RegionID()))
+	}
+	select {
+	case ls.outputCh <- event:
+		ls.lastEvent = event
+		return true
+	default:
+		return false
+	}
+}
+
+// outputResolvedTs nonblocking outputs a resolved ts event.
+func (ls *Sorter) outputResolvedTs(rts model.Ts) {
+	ok := ls.output(model.NewResolvedPolymorphicEvent(0, rts))
+	if ok {
+		ls.lastSentResolvedTs = rts
+	}
+}
+
+// outputBufferedResolvedEvents nonblocking output resolved events and
+// resolved ts that are buffered in outputBuffer.
+// It pops outputted events in the buffer and append their key to deleteKeys.
+func (ls *Sorter) outputBufferedResolvedEvents(
+	buffer *outputBuffer, sendResolvedTsHint bool,
+) {
+	hasRemainEvents := false
+	// Index of remaining output events
+	remainIdx := 0
+	// Commit ts of the last outputed events.
+	lastCommitTs := uint64(0)
+	for idx := range buffer.resolvedEvents {
+		event := buffer.resolvedEvents[idx]
+		ok := ls.output(event)
+		if !ok {
+			hasRemainEvents = true
+			break
+		}
+		lastCommitTs = event.CRTs
+
+		// Delete sent events.
+		key := encoding.EncodeKey(ls.uid, ls.tableID, event)
+		buffer.appendDeleteKey(message.Key(key))
+		remainIdx = idx + 1
+	}
+	// Remove outputted events.
+	buffer.shiftResolvedEvents(remainIdx)
+
+	// If all buffered resolved events are sent, send its resolved ts too.
+	if sendResolvedTsHint && lastCommitTs != 0 && !hasRemainEvents {
+		ls.outputResolvedTs(lastCommitTs)
```

Comment:
Is it necessary to handle the case in which the resolved ts is not sent successfully to a full buffered channel?

Reply:
It's ok if channel is full, it will retry on new received events.